src/node_crypto.cc

```diff
@@ -2579,6 +2580,839 @@ static bool IsSupportedAuthenticatedMode(const EVP_CIPHER_CTX* ctx) {
+    return ManagedEVPPKey(pkey.release());
+  } else {
+    CHECK(args[*offset]->IsObject() && allow_key_object);
+    KeyObject* key;
+    ASSIGN_OR_RETURN_UNWRAP(&key, args[*offset].As<Object>(), ManagedEVPPKey());
+    CHECK_EQ(key->GetKeyType(), kKeyTypePublic);
+    (*offset) += 3;
+    return key->GetAsymmetricKey();
+  }
+}
+
+static NonCopyableMaybe<PrivateKeyEncodingConfig> GetPrivateKeyEncodingFromJS(
+    const FunctionCallbackInfo<Value>& args,
+    unsigned int* offset,
+    KeyEncodingContext context) {
+  Environment* env = Environment::GetCurrent(args);
+
+  PrivateKeyEncodingConfig result;
+  GetKeyFormatAndTypeFromJS(&result, args, offset, context);
+
+  if (result.output_key_object_) {
+    if (context != kKeyContextInput)
+      (*offset)++;
+  } else {
+    bool needs_passphrase = false;
+    if (context != kKeyContextInput) {
+      if (args[*offset]->IsString()) {
+        String::Utf8Value cipher_name(env->isolate(),
+                                      args[*offset].As<String>());
+        result.cipher_ = EVP_get_cipherbyname(*cipher_name);
+        if (result.cipher_ == nullptr) {
+          env->ThrowError("Unknown cipher");
+          return NonCopyableMaybe<PrivateKeyEncodingConfig>();
+        }
+        needs_passphrase = true;
+      } else {
+        CHECK(args[*offset]->IsNullOrUndefined());
+        result.cipher_ = nullptr;
+      }
+      (*offset)++;
+    }
+
+    if (args[*offset]->IsString() || Buffer::HasInstance(args[*offset])) {
+      CHECK_IMPLIES(context != kKeyContextInput, result.cipher_ != nullptr);
+
+      result.passphrase_ = ByteSource::NullTerminatedCopy(env, args[*offset]);
+    } else {
+      CHECK(args[*offset]->IsNullOrUndefined() && !needs_passphrase);
+    }
+  }
+
+  (*offset)++;
+  return NonCopyableMaybe<PrivateKeyEncodingConfig>(std::move(result));
+}
+
+static ManagedEVPPKey GetPrivateKeyFromJS(
+    const FunctionCallbackInfo<Value>& args,
+    unsigned int* offset,
+    bool allow_key_object) {
+  if (args[*offset]->IsString() || Buffer::HasInstance(args[*offset])) {
+    Environment* env = Environment::GetCurrent(args);
+    ByteSource key = ByteSource::FromStringOrBuffer(env, args[(*offset)++]);
+    NonCopyableMaybe<PrivateKeyEncodingConfig> config =
+        GetPrivateKeyEncodingFromJS(args, offset, kKeyContextInput);
+    if (config.IsEmpty())
+      return ManagedEVPPKey();
+    EVPKeyPointer pkey =
+        ParsePrivateKey(config.Release(), key.get(), key.size());
+    if (!pkey)
+      ThrowCryptoError(env, ERR_get_error(), "Failed to read private key");
+    return ManagedEVPPKey(pkey.release());
+  } else {
+    CHECK(args[*offset]->IsObject() && allow_key_object);
+    KeyObject* key;
+    ASSIGN_OR_RETURN_UNWRAP(&key, args[*offset].As<Object>(), ManagedEVPPKey());
+    CHECK_EQ(key->GetKeyType(), kKeyTypePrivate);
+    (*offset) += 4;
+    return key->GetAsymmetricKey();
+  }
+}
+
+static bool IsRSAPrivateKey(const unsigned char* data, size_t size) {
```

Comment:
DER parsing looks good, you are being very careful about fuzzed values and overflows. I'm surprised this is necessary, though I like it, because I didn't see any need to guess in the public js API.